### PR TITLE
#1245 Fix handling of errors during adding of claims

### PR DESCRIPTION
**Pull request by krzyk**
#1245 
- added a check in StkSafe that won't send out `Error` claim if the exception is thrown from `Claims.add` method
- removed todo

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1296](https://github.com/zerocracy/farm/issues/1296) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1296](https://github.com/zerocracy/farm/issues/1296) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/1811052061); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/zerocracy/farm/StkSafe.java [hunk 1]
```diff
@@ -147,4 +129,44 @@ public void process(final Project project,
...
+                .postTo(project);
+        }
+    }
+
+    /**
+     * Is this exception from a Claims.add method?
+     * @param throwable Throwable to check
+     * @return True if from Claims.add
+     */
+    private static boolean fromClaimsAdd(final Throwable throwable) {
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/farm/StkSafe.java]: @krzyk I would rename this method to ``fromAddedClaim``, makes more sense.

File: src/main/java/com/zerocracy/farm/StkSafe.java [hunk 2]
```diff
@@ -147,4 +129,44 @@ public void process(final Project project,
...
+        return throwable.getStackTrace() != null
+            && Arrays.stream(throwable.getStackTrace())
+            .anyMatch(
+                trace -> Objects.equals(
+                    trace.getClassName(), Claims.class.getCanonicalName()
+                ) && Objects.equals(trace.getMethodName(), "add")
+            );
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/farm/StkSafe.java]: @krzyk Is there no better way to do this check? This seems like the most primitive option :(

**Comment by amihaiemil** [comment]
@g4s8 FYI, the class StkSafe is quite monstruous. I don't particularly like the solution from this PR (2 private static methods), but it is the best at the moment. Maybe we need a ``todo`` to break down the class into decorators, or something?

**Comment by krzyk** [comment]
@amihaiemil please check the updated code

**Comment by amihaiemil** [comment]
@krzyk Looks better, thanks :)

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1296](https://github.com/zerocracy/farm/issues/1296) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Three issues were found during code review, [one](https://github.com/zerocracy/farm/pull/1296#discussion_r199355830) of which is cosmetic concern.
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by amihaiemil** [comment]
@ypshenychka confirmed :) I also noticed the class is too big and I'm going to think of a way to refactor it and open a ticket later

**Comment by ypshenychka** [comment]
@amihaiemil I see. However this issue is not addressed by PR's performer, so I can't count it towards code review issue. Thanks!

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
> @0crat quality acceptable ([here](https://github.com/zerocracy/farm/issues/1296#issuecomment-402042597))

@ypshenychka There is no quality review for [#1296](https://github.com/zerocracy/farm/issues/1296), no performer

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**